### PR TITLE
Fix Travis & Travis Build Stages

### PR DESCRIPTION
**Pull request by DirectXMan12**
This fixes our Travis auto-deployment mechnanisms, and switches
over to making use of Travis build stages along the way.

There are now three stages:

1. verify: this *just* runs the verification suite, which currently
   consists of the flake8 tests.  The verification is run under
   both Python 2 and Python 3, using the fedora:rawhide image

2. test: this runs all of non-verification tests, under the full
   test matrix we were using before

3. deploy: this runs our deployements, as before

The deploy phase now precompiles the sdist, and asks for that to be
uploaded instead of regenerating the code before upload.  This is a bit
of a hack.

Finally, the actual Travis scripts themselves were refactored into
a series of smaller files and utility functions, which makes the code
a bit more readable in some cases, and allows individual functions
to be called separately (useful in testing stuff out, and in creating
new build stages).

This PR also fixes a verification failure, and switches the `tox` verification
over to using the lib-verify functions created for Travis.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@frozencemetery 🎉 

this should work, I believe.  There's one last thing to verify, once you review this.

Also, I have some docker-related optimizations in the wings that go on top of this code, but I wasn't able to quite finish them tonight (the gist is that we'd build Docker images on a nightly basis using Travis cron jobs for each item in our build matrix, so that we only have to install the project dependencies in the docker container, which could speed things up a bit)

File: tox.ini [hunk 1]
```diff
@@ -9,10 +9,9 @@ envlist = py27,py33,py34,py35
 [testenv]
 # NB(sross): disabling E225,E226,E227,E901 make pep8 think Cython is ok
 # NB(sross): disable E402 because of the GSSAPI="blah" lines needed by Cython
```
Comment (frozencemetery) [inline tox.ini]: These comments should be moved to where they live now.  (They're duplicated there, so I think they should go away entirely.)

File: .travis/lib-setup.sh [hunk 2]
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
```
Comment (frozencemetery) [inline .travis/lib-setup.sh]: shebang with `-ex` so we see failures and trace from this.

File: .travis/lib-setup.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    if [ -f /etc/fedora-release ]; then
+        setup::fedora::install
+    else
+        setup::centos::install
+    fi
+}
+
+setup::install() {
+    if [ -f /etc/debian_version ]; then
+        setup::debian::install
```
Comment (frozencemetery) [inline .travis/lib-setup.sh]: I'm not sold on these being their own functions, but if you are then it's fine.

File: .travis/lib-util.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Comment (frozencemetery) [inline .travis/lib-util.sh]: Shebang with -ex.

File: .travis/lib-verify.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Comment (frozencemetery) [inline .travis/lib-verify.sh]: Shebang with -ex.

File: .travis/lib-verify.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+verify::flake8() {
```
Comment (frozencemetery) [inline .travis/lib-verify.sh]: Would it be better to move all the lib-*.sh functions into one file? I'm not sold on having this many files, especially when a few of them only have a single function in them.

File: tox.ini (hunk 1 above)
Comment (DirectXMan12) [reply tox.ini]: ack, I just forgot to remove them.

File: .travis/lib-verify.sh (hunk 6 above)
Comment (DirectXMan12) [reply .travis/lib-verify.sh]: I just like splitting things up logically (I almost prefer smaller bash files anyway -- they're easier to manage)

**Pull request merged**